draft-duke-quic-load-balancers.md

```diff
@@ -505,6 +518,208 @@ reserved for config rotation {{config-rotation}}.
+the service's retry tokens. It does not require that all traffic pass through
+the Retry service, so servers MAY send Retry packets.
+
+Tokens the service delivers in Retry packets MUST use the following format. This
+format is the cleartext version. On the wire, these fields are encrypted using
+an AES-ECB cipher and the token key. If the token is not a multiple of 16
+octets, the last block is padded with zeroes.
+
+~~~~~
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|1|  ODCIL    |   Original Destination Connection ID (0..160)   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                             ...                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                               |      
++                                                               +
+|                                                               |
++               Client IP Address ^ date-time (128)             +
+|                                                               |
++                                                               +
+|                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                      rest of date-time (32)                   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~~~
+{: #ss-retry-token-format title="Cleartext format of shared-state retry tokens"}
+
+Retry tokens have the following fields:
+
+ODCIL: The original destination connection ID length.
+
+Original Destination Connection ID: This is copied from the field in the client
+Initial packet.
+
+Client IP Address ^ date-time: This field is the XOR of two values: the client
+IP address and the date-time, as described below.
+
+rest of date-time: The date-time string is a total of 20 octets. The last four
+are in this field, with no XOR operation.
+
+The client IP address is 16 octets. If an IPv4 address, the last 12 octets are
+zeroes.
+
+The format of date-time is described in Section 5.6 of {{!RFC3339}}. This ASCII
+field MUST use the "Z" character for time-offset.
+
+As two fields are XORed in creating the token, a host needs to XOR the IP
+address from the IP header to extract the full date-time and check if it is
+within expiration limits.
+
+### Service Requirements
+
+The service MUST share a "token key" with all supported servers.
+
+When in active mode, the service MUST generate Retry tokens with the format
+described above when it receives a client Initial packet with no token.
+
+In either mode, the service MAY trial-decrypt the tokens in incoming client
+Initial packets. If the first octet of the token is all zeroes, the service MUST forward the packet to the server. Otherwise, the token is ostensibly
```

Comment:
I don't understand the all zeroes part. Before or after decryption? Can you elaborate?

Reply:
after decryption. the point is that randomly generated tokens are not a good way to get through the retry token. Though, I guess opening a valid connection and getting NEW_TOKEN frame is a way to get through.